link_test.go

```diff
@@ -664,3 +667,151 @@ func TestLinkStats64UnmarshalBinary(t *testing.T) {
 		})
 	}
 }
+
+func TestLinkXDP(t *testing.T) {
+	// BPF compilation requires a high RLIMIT_MEMLOCK
+	n := uint64(1024 * 1024 * 10)
+	err := unix.Setrlimit(unix.RLIMIT_MEMLOCK, &unix.Rlimit{Cur: n, Max: n})
+	if err != nil {
+		t.Fatalf("failed to increase RLIMIT_MEMLOCK: %s", err)
+	}
+
+	// establish a netlink connections
+	conn, err := Dial(nil)
+	if err != nil {
+		t.Fatalf("failed to establish netlink socket")
+	}
+	defer conn.Close()
+
+	// construct dummy interface to test XDP program against
+	if err := conn.Link.New(&LinkMessage{
+		Family: unix.AF_UNSPEC,
+		Index:  1001,
+		Flags:  unix.IFF_UP,
+		Attributes: &LinkAttributes{
+			Name: "dummy0",
+			Info: &LinkInfo{Kind: "dummy"},
+		},
+	}); err != nil {
+		t.Fatalf("failed to construct dummy interface: %s", err)
+	}
+
+	// get info for the dummy interface
+	interf, err := conn.Link.Get(1001)
+	if err != nil {
+		conn.Link.Delete(interf.Index)
+		t.Fatalf("could not get dummy interface info: %s", err)
+	}
+
+	// compile a BPF test program. If it fails error out of the tests
+	// and clean up dummy interface.
+	prog, err := ebpf.NewProgram(&ebpf.ProgramSpec{
+		Type: ebpf.XDP,
+		Instructions: asm.Instructions{
+			asm.LoadImm(asm.R0, int64(2), asm.DWord),
+			asm.Return(),
+		},
+		License: "MIT",
+	})
+	if err != nil {
+		conn.Link.Delete(interf.Index)
+		t.Fatalf("failed to compile test bpf program: %s", err)
+	}
+
+	prog2, err := ebpf.NewProgram(&ebpf.ProgramSpec{
+		Type: ebpf.XDP,
+		Instructions: asm.Instructions{
+			asm.LoadImm(asm.R0, int64(4), asm.DWord),
+			asm.Return(),
+		},
+		License: "MIT",
+	})
+	if err != nil {
+		conn.Link.Delete(interf.Index)
+		t.Fatalf("failed to compile test bpf program: %s", err)
+	}
+	// Use the file descriptor of the program
+	progFD := int32(prog.FD())
+	progFD2 := int32(prog2.FD())
+
+	message := LinkMessage{
+		Family:     unix.AF_UNSPEC,
+		Index:      interf.Index,
+		Attributes: &LinkAttributes{},
+	}
+
+	tests := []struct {
+		name     string
+		xdp      *LinkXDP
+		succes   bool
+		attached uint8
+	}{
+		{
+			"load XDP program",
+			&LinkXDP{
+				FD:    progFD,
+				Flags: linux.XDP_FLAGS_SKB_MODE,
+			},
+			true,
+			2,
+		},
+		{
+			"load XDP program 2 wrong expected FD",
+			&LinkXDP{
+				FD:         progFD2,
+				ExpectedFD: progFD2,
+				Flags:      linux.XDP_FLAGS_SKB_MODE | linux.XDP_FLAGS_REPLACE,
+			},
+			false,
+			2,
+		},
+		{
+			"load XDP program 2",
+			&LinkXDP{
+				FD:         progFD2,
+				ExpectedFD: progFD,
+				Flags:      linux.XDP_FLAGS_SKB_MODE | linux.XDP_FLAGS_REPLACE,
+			},
+			true,
+			2,
+		},
+		{
+			"clear XDP program",
+			&LinkXDP{
+				FD:    -1,
+				Flags: linux.XDP_FLAGS_SKB_MODE,
+			},
+			true,
+			0,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			message.Attributes.XDP = tt.xdp
+			err := conn.Link.Set(&message)
+			if err != nil && tt.succes {
+				t.Errorf("failed to send netlink message: %s", err)
+			}
+		})
```

Comment:
The subtest should encompass this entire loop body rather than just the Set call.